File: tests/cuda/python/test_elementwise.py
"""Elementwise CUDA binding tests — aligned with C++ test_elementwise.cpp.

Run with:
    PYTHONPATH=bindings/python:build/bindings/python
    LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda
    python -m pytest tests/cuda/python/test_elementwise.py -v
"""

import sys
import os
import pytest

_root = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "..")
sys.path.insert(0, os.path.join(_root, "bindings", "python"))
sys.path.insert(0, os.path.join(_root, "build", "bindings", "python"))

try:
    import insight as ins
    import numpy as np
except ImportError:
    pytest.skip("Insight or NumPy not available", allow_module_level=True)

try:
    ins.init()
    if ins.device_count() == 0:
        pytest.skip("No GPU available", allow_module_level=True)
except Exception:
    pytest.skip("GPU backend not available", allow_module_level=True)

GPU = ins.GPUPlace(0)
CPU = ins.CPUPlace()


def to_gpu(np_arr, dtype=None):
    if dtype is not None:
        np_arr = np_arr.astype(dtype)
    return ins.from_numpy(np_arr).to(GPU)


def to_numpy(gpu_arr):
    return gpu_arr.to(CPU).numpy()


class TestElementwiseCUDA:
    """Elementwise binary operations — CUDA backend."""

    def test_add(self):
        a_np = np.array([1.0, 2.0, 3.0, 4.0, 5.0], dtype=np.float64)
        b_np = np.array([5.0, 4.0, 3.0, 2.0, 1.0], dtype=np.float64)
        result = ins.add(to_gpu(a_np), to_gpu(b_np))
        np.testing.assert_allclose(to_numpy(result), a_np + b_np, rtol=1e-10)

    def test_sub(self):
        a_np = np.array([10.0, 20.0, 30.0], dtype=np.float64)
        b_np = np.array([1.0, 2.0, 3.0], dtype=np.float64)
        result = ins.sub(to_gpu(a_np), to_gpu(b_np))
        np.testing.assert_allclose(to_numpy(result), a_np - b_np, rtol=1e-10)

    def test_mul(self):
        a_np = np.array([2.0, 3.0, 4.0], dtype=np.float64)
        b_np = np.array([5.0, 6.0, 7.0], dtype=np.float64)
        result = ins.mul(to_gpu(a_np), to_gpu(b_np))
        np.testing.assert_allclose(to_numpy(result), a_np * b_np, rtol=1e-10)

    def test_div(self):
        a_np = np.array([10.0, 20.0, 30.0], dtype=np.float64)
        b_np = np.array([2.0, 4.0, 5.0], dtype=np.float64)
        result = ins.div(to_gpu(a_np), to_gpu(b_np))
        np.testing.assert_allclose(to_numpy(result), a_np / b_np, rtol=1e-10)

    def test_pow(self):
        a_np = np.array([2.0, 3.0, 4.0], dtype=np.float64)
        b_np = np.array([3.0, 2.0, 0.5], dtype=np.float64)
        result = ins.pow(to_gpu(a_np), to_gpu(b_np))
        np.testing.assert_allclose(to_numpy(result), a_np**b_np, rtol=1e-6)

    def test_mod(self):
        a_np = np.array([10.0, 7.0, 5.0], dtype=np.float64)
        b_np = np.array([3.0, 2.0, 3.0], dtype=np.float64)
        result = ins.mod(to_gpu(a_np), to_gpu(b_np))
        np.testing.assert_allclose(to_numpy(result), a_np % b_np, rtol=1e-10)

    def test_maximum(self):
        a_np = np.array([1.0, 5.0, 3.0], dtype=np.float64)
        b_np = np.array([4.0, 2.0, 6.0], dtype=np.float64)
        result = ins.maximum(to_gpu(a_np), to_gpu(b_np))
        np.testing.assert_allclose(to_numpy(result), np.maximum(a_np, b_np))

    def test_minimum(self):
        a_np = np.array([1.0, 5.0, 3.0], dtype=np.float64)
        b_np = np.array([4.0, 2.0, 6.0], dtype=np.float64)
        result = ins.minimum(to_gpu(a_np), to_gpu(b_np))
        np.testing.assert_allclose(to_numpy(result), np.minimum(a_np, b_np))

    def test_equal(self):
        a = to_gpu(np.array([1, 2, 3], dtype=np.float64))
        b = to_gpu(np.array([1, 0, 3], dtype=np.float64))
        result = ins.equal(a, b)
        np.testing.assert_array_equal(to_numpy(result), np.array([True, False, True]))

    def test_not_equal(self):
        a = to_gpu(np.array([1, 2, 3], dtype=np.float64))
        b = to_gpu(np.array([1, 0, 3], dtype=np.float64))
        result = ins.not_equal(a, b)
        np.testing.assert_array_equal(to_numpy(result), np.array([False, True, False]))

    def test_greater(self):
        a = to_gpu(np.array([3, 1, 5], dtype=np.float64))
        b = to_gpu(np.array([2, 4, 5], dtype=np.float64))
        result = ins.greater(a, b)
        np.testing.assert_array_equal(to_numpy(result), np.array([True, False, False]))

    def test_less(self):
        a = to_gpu(np.array([1, 5, 3], dtype=np.float64))
        b = to_gpu(np.array([2, 4, 5], dtype=np.float64))
        result = ins.less(a, b)
        np.testing.assert_array_equal(to_numpy(result), np.array([True, False, True]))

    def test_greater_equal(self):
        a = to_gpu(np.array([3, 1, 5], dtype=np.float64))
        b = to_gpu(np.array([2, 4, 5], dtype=np.float64))
        result = ins.greater_equal(a, b)
        np.testing.assert_array_equal(to_numpy(result), np.array([True, False, True]))

    def test_less_equal(self):
        a = to_gpu(np.array([1, 5, 3], dtype=np.float64))
        b = to_gpu(np.array([2, 4, 5], dtype=np.float64))
        result = ins.less_equal(a, b)
        np.testing.assert_array_equal(to_numpy(result), np.array([True, False, True]))

    def test_logical_and(self):
        a = to_gpu(np.array([True, True, False], dtype=np.bool_))
        b = to_gpu(np.array([True, False, False], dtype=np.bool_))
        result = ins.logical_and(a, b)
        np.testing.assert_array_equal(to_numpy(result), np.array([True, False, False]))

    def test_logical_or(self):
        a = to_gpu(np.array([True, True, False], dtype=np.bool_))
        b = to_gpu(np.array([True, False, False], dtype=np.bool_))
        result = ins.logical_or(a, b)
        np.testing.assert_array_equal(to_numpy(result), np.array([True, True, False]))

    def test_logical_xor(self):
        a = to_gpu(np.array([True, True, False], dtype=np.bool_))
        b = to_gpu(np.array([True, False, False], dtype=np.bool_))
        result = ins.logical_xor(a, b)
        np.testing.assert_array_equal(to_numpy(result), np.array([False, True, False]))

    def test_logical_not(self):
        a = to_gpu(np.array([True, False, True], dtype=np.bool_))
        result = ins.logical_not(a)
        np.testing.assert_array_equal(to_numpy(result), np.array([False, True, False]))


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/cuda/julia/test_elementwise.jl
# Elementwise CUDA tests — aligned with C++ test_elementwise.cpp
# Run with:
#   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda julia tests/cuda/julia/test_elementwise.jl

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "build", "bindings", "julia"))

using Insight

try
    Insight.load_backend("gpu")
    Insight.set_device(Insight.GPUPlace(0))
catch e
    println("SKIP: GPU backend not available: $e")
    exit(0)
end

passed = 0; failed = 0
function check(name, cond)
    global passed, failed
    if cond; passed += 1; else; failed += 1; println("FAIL: $name"); end
end

println("=== Elementwise CUDA ===")

a = Insight.to_array(Insight.from_data([1.0, 2.0, 3.0, 4.0, 5.0]), Insight.GPUPlace(0))
b = Insight.to_array(Insight.from_data([5.0, 4.0, 3.0, 2.0, 1.0]), Insight.GPUPlace(0))

# add
c = Insight.add(a, b)
check("add", Insight.numel(c) == 5)

# sub
c = Insight.sub(a, b)
check("sub", Insight.numel(c) == 5)

# mul
c = Insight.mul(a, b)
check("mul", Insight.numel(c) == 5)

# div
c = Insight.div(a, b)
check("div", Insight.numel(c) == 5)

# pow
a2 = Insight.to_array(Insight.from_data([2.0, 3.0, 4.0]), Insight.GPUPlace(0))
b2 = Insight.to_array(Insight.from_data([3.0, 2.0, 0.5]), Insight.GPUPlace(0))
c = Insight.pow(a2, b2)
check("pow", Insight.numel(c) == 3)

# add operator
c = a + b
check("add_op", Insight.numel(c) == 5)

# sub operator
c = a - b
check("sub_op", Insight.numel(c) == 5)

# mul operator
c = a * b
check("mul_op", Insight.numel(c) == 5)

# div operator
c = a / b
check("div_op", Insight.numel(c) == 5)

# negation
c = -a
check("neg", Insight.numel(c) == 5)

# equal
a3 = Insight.to_array(Insight.from_data([1.0, 2.0, 3.0]), Insight.GPUPlace(0))
b3 = Insight.to_array(Insight.from_data([1.0, 0.0, 3.0]), Insight.GPUPlace(0))
c = Insight.equal(a3, b3)
check("equal", Insight.numel(c) == 3)

# greater
a4 = Insight.to_array(Insight.from_data([3.0, 1.0, 5.0]), Insight.GPUPlace(0))
b4 = Insight.to_array(Insight.from_data([2.0, 4.0, 5.0]), Insight.GPUPlace(0))
c = Insight.greater(a4, b4)
check("greater", Insight.numel(c) == 3)

# less
c = Insight.less(a4, b4)
check("less", Insight.numel(c) == 3)

# logical_and
a_bool = Insight.to_array(Insight.from_data([1, 1, 0], Insight.bool), Insight.GPUPlace(0))
b_bool = Insight.to_array(Insight.from_data([1, 0, 0], Insight.bool), Insight.GPUPlace(0))
c = Insight.logical_and(a_bool, b_bool)
check("logical_and", Insight.numel(c) == 3)

# logical_or
c = Insight.logical_or(a_bool, b_bool)
check("logical_or", Insight.numel(c) == 3)

# logical_not
c = Insight.logical_not(a_bool)
check("logical_not", Insight.numel(c) == 3)

println("\n" * "="^40)
println("Results: $passed passed, $failed failed")
if failed > 0; exit(1); end
